File: src/main/java/net/atos/zac/webdav/WebdavHelper.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.webdav;

import static java.lang.String.format;
import static org.apache.commons.io.FilenameUtils.getExtension;

import java.net.URI;
import java.util.Collections;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import jakarta.ws.rs.core.UriInfo;

import org.apache.commons.collections4.map.LRUMap;

import net.atos.client.zgw.drc.DrcClientService;
import net.atos.zac.util.MediaTypes;
import nl.info.client.zgw.drc.model.generated.EnkelvoudigInformatieObject;
import nl.info.zac.authentication.LoggedInUser;

@Singleton
public class WebdavHelper {
    public static final String FOLDER = "folder";
    private static final Set<String> WEBDAV_WORD = Set.of(
            MediaTypes.Application.MS_WORD.getMediaType(),
            MediaTypes.Application.MS_WORD_OPEN_XML.getMediaType()
    );
    private static final Set<String> WEBDAV_POWERPOINT = Set.of(
            MediaTypes.Application.MS_POWER_POINT.getMediaType(),
            MediaTypes.Application.MS_POWER_POINT_OPEN_XML.getMediaType()
    );
    private static final Set<String> WEBDAV_EXCEL = Set.of(
            MediaTypes.Application.MS_EXCEL.getMediaType(),
            MediaTypes.Application.MS_EXCEL_OPEN_XML.getMediaType()
    );

    @Inject
    private DrcClientService drcClientService;

    @Inject
    private Instance<LoggedInUser> loggedInUserInstance;

    private final Map<String, Gegevens> tokenMap = Collections.synchronizedMap(new LRUMap<>(1000));

    public URI createRedirectURL(final UUID enkelvoudigInformatieobjectUUID, final UriInfo uriInfo) {
        final EnkelvoudigInformatieObject enkelvoudigInformatieobject = drcClientService.readEnkelvoudigInformatieobject(
                enkelvoudigInformatieobjectUUID);
        final String scheme = format(
                "%s:%s",
                getWebDAVApp(enkelvoudigInformatieobject.getFormaat()),
                uriInfo.getBaseUri().getScheme()
        );
        final String token = format(
                "%s.%s",
                createToken(enkelvoudigInformatieobjectUUID),
                getExtension(enkelvoudigInformatieobject.getBestandsnaam())
        );
        return uriInfo.getBaseUriBuilder().scheme(scheme).replacePath("webdav/folder/{token}").build(token);
    }

    public Gegevens readGegevens(final String token) {
        if (tokenMap.containsKey(token)) {
            return tokenMap.get(token);
        } else {
            throw new RuntimeException("WebDAV token does not exist (anymore).");
        }
    }

    private String createToken(final UUID enkelvoudigInformatieobjectUUID) {
        final String token = UUID.randomUUID().toString();
        tokenMap.put(token, new Gegevens(enkelvoudigInformatieobjectUUID, loggedInUserInstance.get()));
        return token;
    }

    private String getWebDAVApp(final String formaat) {
        if (WEBDAV_WORD.contains(formaat)) {
            return "ms-word";
        }
        if (WEBDAV_EXCEL.contains(formaat)) {
            return "ms-excel";
        }
        if (WEBDAV_POWERPOINT.contains(formaat)) {
            return "ms-powerpoint";
        }
        return null;
    }

    public record Gegevens(UUID enkelvoudigInformatieibjectUUID, LoggedInUser loggedInUser) {
    }
}

File: src/main/kotlin/nl/info/zac/webdav/WebdavHelper.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.webdav

import jakarta.enterprise.inject.Instance
import jakarta.inject.Inject
import jakarta.inject.Singleton
import jakarta.ws.rs.core.UriInfo
import net.atos.zac.util.MediaTypes
import nl.info.client.zgw.drc.DrcClientService
import nl.info.webdav.exceptions.WebdavException
import nl.info.zac.authentication.LoggedInUser
import nl.info.zac.util.NoArgConstructor
import org.apache.commons.collections4.map.LRUMap
import org.apache.commons.io.FilenameUtils.getExtension
import java.net.URI
import java.util.Collections
import java.util.UUID

@Singleton
@NoArgConstructor
class WebdavHelper @Inject constructor(
    private val drcClientService: DrcClientService,
    private val loggedInUserInstance: Instance<LoggedInUser>
) {
    companion object {
        const val FOLDER = "folder"
        const val TOKEN_MAP_MAX_SIZE = 1000
        private val WEBDAV_WORD = setOf(
            MediaTypes.Application.MS_WORD.mediaType,
            MediaTypes.Application.MS_WORD_OPEN_XML.mediaType
        )
        private val WEBDAV_POWERPOINT = setOf(
            MediaTypes.Application.MS_POWER_POINT.mediaType,
            MediaTypes.Application.MS_POWER_POINT_OPEN_XML.mediaType
        )
        private val WEBDAV_EXCEL = setOf(
            MediaTypes.Application.MS_EXCEL.mediaType,
            MediaTypes.Application.MS_EXCEL_OPEN_XML.mediaType
        )
    }

    private val tokenMap: MutableMap<String, WebdavTokenData> = Collections.synchronizedMap(LRUMap(TOKEN_MAP_MAX_SIZE))

    fun createRedirectURL(enkelvoudigInformatieobjectUUID: UUID, uriInfo: UriInfo): URI {
        val enkelvoudigInformatieobject = drcClientService.readEnkelvoudigInformatieobject(
            enkelvoudigInformatieobjectUUID
        )
        val scheme = "${getWebdavApp(enkelvoudigInformatieobject.formaat)}:${uriInfo.baseUri.scheme}"
        val token = "${createToken(
            enkelvoudigInformatieobjectUUID
        )}.${getExtension(enkelvoudigInformatieobject.bestandsnaam)}"
        return uriInfo.baseUriBuilder.scheme(scheme).replacePath("webdav/folder/{token}").build(token)
    }

    fun readWebdavTokenData(token: String): WebdavTokenData =
        tokenMap[token] ?: throw WebdavException("WebDAV token does not exist (anymore).")

    private fun createToken(enkelvoudigInformatieobjectUUID: UUID): String {
        val token = UUID.randomUUID().toString()
        tokenMap[token] = WebdavTokenData(
            enkelvoudigInformatieobjectUUID = enkelvoudigInformatieobjectUUID,
            loggedInUser = loggedInUserInstance.get()
        )
        return token
    }

    private fun getWebdavApp(formaat: String): String? = when {
        WEBDAV_WORD.contains(formaat) -> "ms-word"
        WEBDAV_EXCEL.contains(formaat) -> "ms-excel"
        WEBDAV_POWERPOINT.contains(formaat) -> "ms-powerpoint"
        else -> null
    }

    data class WebdavTokenData(val enkelvoudigInformatieobjectUUID: UUID, val loggedInUser: LoggedInUser)
}
